File: Data Structures/Arrays/Hashing/LeetCode/49GroupAnagram/49GroupAnagram.py
from typing import List

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagrams = {}  # Dictionary to store grouped anagrams

        for word in strs:
            # Sort the word to use as a key
            sorted_word = ''.join(sorted(word))
            
            # Group words by their sorted key
            if sorted_word not in anagrams:
                anagrams[sorted_word] = []
            anagrams[sorted_word].append(word)

        # Return grouped anagrams as a list of lists
        return list(anagrams.values())
    
    """
    Explanation:
Initialize a Dictionary:

anagrams is a dictionary where the key is the sorted version of a word, and the value is a list of words (anagrams).
Iterate Over Input Strings:

For each word in strs, calculate its sorted version:
''.join(sorted(word)): Sorts the characters in the word and converts them back to a string.
Group Words by Sorted Key:

If the sorted version of the word is not a key in the dictionary, add it with an empty list as the value.
Append the current word to the list corresponding to its sorted key.
Return Grouped Anagrams:

Use list(anagrams.values()) to return all the grouped lists of anagrams.
Complexity:
Time Complexity:

Sorting each word: 
𝑂(𝑘logZ)
O(klogk), where 𝑘
k is the average length of words.
Iterating over 𝑛n words: 𝑂(𝑛⋅𝑘log𝑘)
O(n⋅klogk).
Total: 𝑂(𝑛⋅𝑘log⁡𝑘)
O(n⋅klogk).
Space Complexity:
The dictionary anagrams stores all the words, so 𝑂(𝑛⋅𝑘)
O(n⋅k) space for the keys and values.    
    """

File: Data Structures/Arrays/Hashing/LeetCode/49GroupAnagram/49GroupAnagram.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        // Dictionary to store sorted word as key and a list of anagrams as values
        unordered_map<string, vector<string>> anagrams;

        // Iterate through each word in the input list
        for (string& word : strs) {
            // Sort the word to use as a key
            string sorted_word = word;
            sort(sorted_word.begin(), sorted_word.end());

            // Group the anagrams by their sorted key
            anagrams[sorted_word].push_back(word);
        }

        // Prepare the result: return all values from the map
        vector<vector<string>> result;
        for (auto& pair : anagrams) {
            result.push_back(pair.second);
        }

        return result;
    }
};
